biz.aQute.bndlib/src/aQute/bnd/plugin/jpms/JPMSModuleInfoPlugin.java

```diff
@@ -0,0 +1,397 @@
+	private static final String		INTERNAL_MODULE_DIRECTIVE	= "-internal-module:";
+
+	@Override
+	public void verify(final Analyzer analyzer) throws Exception {
+		String moduleProperty = analyzer.getProperty(JPMS_MODULE_INFO);
+		if (moduleProperty == null)
+			return;
+
+		Parameters provideCapabilities = new Parameters(analyzer.getJar()
+			.getManifest()
+			.getMainAttributes()
+			.getValue(Constants.PROVIDE_CAPABILITY));
+		Parameters requireCapabilities = new Parameters(analyzer.getJar()
+			.getManifest()
+			.getMainAttributes()
+			.getValue(Constants.REQUIRE_CAPABILITY));
+
+		if ("".equals(moduleProperty)) {
+			String name = name(analyzer);
+			int access = access(requireCapabilities);
+			String version = analyzer.getVersion();
+
+			moduleProperty = String.format("%s;access=%s;version=%s", name, access, version);
+		}
+
+		Parameters moduleParameters = OSGiHeader.parseHeader(moduleProperty);
+		if (moduleParameters.size() == 0)
+			return;
+
+		if (moduleParameters.size() > 1)
+			throw new IllegalArgumentException("Only one -module instruction is allowed:" + moduleParameters);
+
+		Entry<String, Attrs> moduleInstructions = moduleParameters.entrySet()
+			.iterator()
+			.next();
+
+		Packages index = new Packages();
+
+		// Index the whole class path
+		for (Jar jar : analyzer.getClasspath()) {
+			String moduleName = getModuleName(analyzer, jar);
+			String moduleVersion = jar.getModuleVersion();
+			jar.getDirectories()
+				.entrySet()
+				.stream()
+				.filter(entry -> (entry.getValue() != null) && !entry.getValue()
+					.isEmpty()
+					&& entry.getKey()
+						.length() > 0
+					&& !entry.getKey()
+						.startsWith("META-INF")
+					&& !entry.getKey()
+						.startsWith("OSGI-INF")
+					&& !entry.getKey()
+						.startsWith("OSGI-OPT")
+					&& !entry.getKey()
+						.startsWith("WEB-INF"))
+				.forEach(entry -> {
+					PackageRef ref = analyzer.getPackageRef(entry.getKey());
+					Attrs attrs = new Attrs();
+					if (moduleName != null) {
+						attrs.put(INTERNAL_MODULE_DIRECTIVE, moduleName);
+					}
+					if (moduleVersion != null) {
+						attrs.put(INTERNAL_MODULE_VERSION_DIRECTIVE, moduleVersion);
+					}
+					index.put(ref, attrs);
+				});
+		}
+
+		ModuleInfoWriter writer = nameAccessAndVersion(moduleInstructions, requireCapabilities, analyzer);
+
+		requires(moduleInstructions, analyzer, index, writer);
+		exportPackages(analyzer, writer);
+		openPackages(analyzer, writer);
+		serviceLoaderProviders(provideCapabilities, analyzer, writer);
+		serviceLoaderUses(requireCapabilities, analyzer, writer);
+		mainClass(analyzer, writer);
+
+		// TODO use annotations to store other header info???
+		// AnnotationVisitor visitAnnotation = classWriter.visitAnnotation(...,
+		// false);
+
+		ByteBufferDataOutput bbout = new ByteBufferDataOutput();
+		writer.write(bbout);
+
+		analyzer.getJar()
+			.putResource(MODULE_INFO_CLASS,
+				new EmbeddedResource(bbout.toByteArray(), analyzer.lastModified()));
```

Comment:
Use `new EmbeddedResource(bbout.toByteBuffer(),...` to avoid creating an array and then wrapping that array in a new ByteBuffer.